atlasaction/action_test.go

```diff
@@ -87,6 +93,234 @@ func TestMigrateApply(t *testing.T) {
+		tt.setInput("tag", "invalid-character@")
+		err := MigratePush(context.Background(), tt.cli, tt.act)
+		require.ErrorContains(t, err, `tag must be lowercase alphanumeric`)
+	})
+	t.Run("tag", func(t *testing.T) {
+		tt := newHttpT(t, nil)
+		tt.setupConfigWithLogin(t)
+		tt.setInput("dir", "file://testdata/migrations")
+		tt.setInput("dir-name", "test-dir")
+		tt.setInput("dev-url", "sqlite://file?mode=memory")
+		tt.setInput("tag", "valid-tag-123")
+		err := MigratePush(context.Background(), tt.cli, tt.act)
+		require.NoError(t, err)
+	})
+	t.Run("e2e test", func(t *testing.T) {
+		var syncDirCalls, pushDirCalls atomicInt
+
+		tt := newHttpT(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			query := graphQLQuery{}
+			err := json.NewDecoder(r.Body).Decode(&query)
+			require.NoError(t, err)
+			switch {
+			case strings.Contains(query.Query, "syncDir"):
+				syncDirCalls.inc()
+				require.NoError(t, json.Unmarshal(query.Variables, &query.SyncDir))
+				require.Equal(t, "test-dir", query.SyncDir.Slug)
+				expected := &ContextInput{
+					Repo:   "repository",
+					Path:   "file://testdata/migrations",
+					Branch: "testing-branch",
+					Commit: "sha1234",
+					URL:    "",
+				}
+				require.Equal(t, expected, query.SyncDir.Context)
+			case strings.Contains(query.Query, "pushDir"):
+				pushDirCalls.inc()
+				require.NoError(t, json.Unmarshal(query.Variables, &query.PushDir))
+				require.Equal(t, query.PushDir.Tag, "sha1234")
+				require.Equal(t, query.PushDir.Slug, "test-dir")
+				fmt.Fprint(w, `{"data":{"pushDir":{"url":"https://some-org.atlasgo.cloud/dirs/314159/tags/12345"}}}`)
+			}
+		}))
+		tt.setupConfigWithLogin(t)
+		tt.setInput("dir", "file://testdata/migrations")
+		tt.setInput("dir-name", "test-dir")
+		tt.setInput("dev-url", "sqlite://file?mode=memory")
+		tt.env["GITHUB_REPOSITORY"] = "repository"
+		tt.env["GITHUB_REF_NAME"] = "testing-branch"
+		tt.env["GITHUB_SHA"] = "sha1234"
+		err := MigratePush(context.Background(), tt.cli, tt.act)
+		require.NoError(t, err)
+		require.Equal(t, syncDirCalls.get(), int32(1))
+		require.Equal(t, pushDirCalls.get(), int32(1))
+		require.NoError(t, err)
+		outputs, _ := tt.outputs()
+		url := outputs["url"]
+		require.Equal(t, "https://some-org.atlasgo.cloud/dirs/314159/tags/12345", url)
+	})
+}
+
+type (
+	httpTest struct {
+		*test
+		srv        *httptest.Server
+		cloudToken string
+	}
+
+	pushDir struct {
+		Slug   string `json:"slug"`
+		Tag    string `json:"tag"`
+		Driver string `json:"driver"`
+		Dir    string `json:"dir"`
+	}
+	syncDir struct {
+		Slug    string        `json:"slug"`
+		Driver  string        `json:"driver"`
+		Dir     string        `json:"dir"`
+		Context *ContextInput `json:"context"`
+	}
+
+	graphQLQuery struct {
+		Query     string          `json:"query"`
+		Variables json.RawMessage `json:"variables"`
+		PushDir   struct {
+			pushDir `json:"input"`
+		}
+		SyncDir struct {
+			syncDir `json:"input"`
+		}
```

Comment:
I suggest to scope these types to specific tests. See this example: https://github.com/ariga/atlas/blob/master/cmd/atlas/internal/cmdapi/migrate_test.go#L790